### PR TITLE
Backport 2.28: Start testing the PSA built-in drivers: hashes

### PR DESCRIPTION
**Pull request by tgonzalezorlandoarm**
## Description

Backport of #7803

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** no (test only)
- [x] **backport** this is the backport
- [x] **tests** provided

## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

### CODE REVIEW

**Comment by tgonzalezorlandoarm** [comment]
A couple of questions:

1. @gilles-peskine-arm I only left your Sign-off and author as these are your commits. I have barely retouched one commit to make it fit to the backport, so I hope this is okay, let me know otherwise.
2. I see that the test does not get generated. I can't find why it's behaving this way, I would appreciate any pointers.

**Comment by gilles-peskine-arm** [comment]
> @gilles-peskine-arm I only left your Sign-off and author as these are your commits. I have barely retouched one commit to make it fit to the backport, so I hope this is okay, let me know otherwise.

That's fine. The signoff line isn't the author line, it's a claim that we have the rights to contribute the code. Between arm employees, we don't really care who signs off: we're all signing off as agents of our employer.

> I see that the test does not get generated. I can't find why it's behaving this way, I would appreciate any pointers.

In 2.28, all generated files are checked into git (because we don't require users to have the necessary tools). When you do something that affects a generated source file, you need to run `tests/scripts/check_generated_files -u` and commit the result.

**Comment by tgonzalezorlandoarm** [comment]
Perfect, thanks for the information @gilles-peskine-arm ! I've just updated the files accordingly

**Comment by gilles-peskine-arm** [comment]
Looks like I merged https://github.com/Mbed-TLS/mbedtls/pull/7803 without waiting for its backport, which is this pull request. That's not ideal, but it's not a big deal either since this is just tests. The original pull request was motivated by improved driver support, which doesn't exist in 2.28: we wanted extra tests to make sure the driver interface supports drivers well. The reason to backport is mostly to facilitate the backports of future tests.

Since this is just tests and not bug fixes, it doesn't matter if we make a release with #7803 but not its backport.

File: tests/scripts/generate_psa_tests.py [hunk 1]
```diff
@@ -189,12 +202,15 @@ def test_cases_for_key_type_key_generation(
         else:
             generate_dependencies = \
                 psa_information.fix_key_pair_dependencies(import_dependencies, 'GENERATE')
-            if kt.name == 'PSA_KEY_TYPE_RSA_KEY_PAIR':
-                generate_dependencies.append("MBEDTLS_GENPRIME")
         for bits in kt.sizes_to_test():
+            if kt.name == 'PSA_KEY_TYPE_RSA_KEY_PAIR':
+                size_dependency = "PSA_VENDOR_RSA_GENERATE_MIN_KEY_BITS <= " +  str(bits)
```
Comment (gilles-peskine-arm) [inline tests/scripts/generate_psa_tests.py:207]: `PSA_VENDOR_RSA_GENERATE_MIN_KEY_BITS` doesn't exist in 2.28.

File: tests/suites/test_suite_psa_crypto_generate_key.generated.data [hunk 2]
```diff
@@ -109,11 +109,11 @@ depends_on:PSA_WANT_KEY_TYPE_RAW_DATA
 generate_key:PSA_KEY_TYPE_RAW_DATA:128:PSA_SUCCESS:
 
 PSA RSA_KEY_PAIR 1024-bit
-depends_on:PSA_WANT_KEY_TYPE_RSA_KEY_PAIR:MBEDTLS_GENPRIME
+depends_on:PSA_WANT_KEY_TYPE_RSA_KEY_PAIR_GENERATE:PSA_VENDOR_RSA_GENERATE_MIN_KEY_BITS <= 1024:DEPENDENCY_NOT_IMPLEMENTED_YET
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_psa_crypto_generate_key.generated.data:112]: `PSA_WANT_KEY_TYPE_RSA_KEY_PAIR_GENERATE` doesn't exist in 2.28. The old dependencies were correct for 2.28.

File: scripts/mbedtls_dev/psa_information.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
+"""Collect information about PSA cryptographic mechanisms.
```
Comment (gilles-peskine-arm) [inline scripts/mbedtls_dev/psa_information.py:1]: The commit "Move PSA information and dependency automation into their own module" should be doing the following:

* Create `psa_information.py` and write some stuff at the top (docstring, copyright notice, imports).
* Move blocks of code from `generate_psa_tests.py` to `psa_information.py`.
* In `generate_psa_tests.py`, add `psa_information.` when referencing functions that have moved to the new module.
* (not in 2.28) update dependencies in makefiles.

b4b5402bfcd4d78c41f98020316295f832add0aa is not doing this correctly: it's adding code from `development`. This commit needs to be reconstructed rather than cherry-picked.

This has consequences, some benign, some not. One bad consequence is that `PSA_WANT_KEY_TYPE_xxx_KEY_PAIR` dependencies have moved to their 3.5 names, which is wrong because those names don't exist in 2.28. This causes some test cases to be never executed.

File: scripts/mbedtls_dev/crypto_knowledge.py [hunk 4]
```diff
@@ -34,7 +34,7 @@ def short_expression(original: str, level: int = 0) -> str:
     unambiguous, but ad hoc way.
     """
     short = original
-    short = re.sub(r'\bPSA_(?:ALG|ECC_FAMILY|KEY_[A-Z]+)_', r'', short)
+    short = re.sub(r'\bPSA_(?:ALG|DH_FAMILY|ECC_FAMILY|KEY_[A-Z]+)_', r'', short)
```
Comment (gilles-peskine-arm) [inline scripts/mbedtls_dev/crypto_knowledge.py:37]: I don't understand: why backport FFDH support to the python scripts? In 2.28, the PSA API doesn't support FFDH.

Also, the commit "Add DH support for psa test generation and update data files" is doing other things.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
The code move wasn't correctly backported.

File: scripts/mbedtls_dev/psa_information.py (hunk 3 above)
Comment (tgonzalezorlandoarm) [reply scripts/mbedtls_dev/psa_information.py:1]: I think I now managed to do the backport properly, sorry for the confusion

File: scripts/mbedtls_dev/psa_information.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                     for symbol in re.findall(r'\bPSA_WANT_\w+\b', line))
+_implemented_dependencies = None #type: Optional[FrozenSet[str]] #pylint: disable=invalid-name
+
+def hack_dependencies_not_implemented(dependencies: List[str]) -> None:
+    global _implemented_dependencies #pylint: disable=global-statement,invalid-name
+    if _implemented_dependencies is None:
+        _implemented_dependencies = \
+            read_implemented_dependencies('include/psa/crypto_config.h')
+    if not all((dep.lstrip('!') in _implemented_dependencies or
+                not dep.lstrip('!').startswith('PSA_WANT'))
```
Comment (gilles-peskine-arm) [inline scripts/mbedtls_dev/psa_information.py:84]: 734d22c03ee68192393ec660b3512a425c17e4aa changed these two lines from
```
    if not all((dep.lstrip('!') in _implemented_dependencies or 'PSA_WANT' not in dep)
```
The new code is better, so that's ok. It would have been better to do this in a separate commit, but never mind now.

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Comment by minosgalanakis** [comment]
This PR looks good code-wise but I think it should be be updating to the newer licence headers.

**Review by minosgalanakis — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM
